src/net/phoenix.rs

```diff
@@ -46,67 +46,117 @@ pub fn spawn(
                     }
+                });
+
+                tracing::info!(addr=%listener.local_addr(), "starting phoenix HTTP service");
+                let (stx, srx) = tokio::sync::oneshot::channel::<()>();
+
+                let http_task = tokio::spawn(
+                    HyperServer::from_tcp(listener.into())?
+                        .serve(make_svc)
+                        .with_graceful_shutdown(async move {
+                            let _ = srx.await;
+                        }),
+                );
+
+                let res = loop {
+                    use eyre::WrapErr as _;
+
+                    tokio::select! {
+                        _ = shutdown_rx.changed() => break Ok::<_, eyre::Error>(()),
+                        result = config_watcher.changed() => if let Err(err) = result {
+                            break Err(err).context("config watcher sender dropped");
+                        },
+                        result = phoenix_watcher.changed() => if let Err(err) = result {
+                            break Err(err).context("phoenix watcher sender dropped");
+                        },
+                    }
+
+                    tracing::trace!("change detected, updating phoenix");
+                    phoenix.add_nodes_from_config(&config);
+                    let nodes = phoenix.ordered_nodes_by_latency();
+                    let mut new_json = serde_json::Map::default();
+
+                    for (identifier, latency) in nodes {
+                        new_json.insert(identifier.to_string(), latency.into());
+                    }
+
+                    json.store(new_json.into());
+                };
+
+                if stx.send(()).is_err() {
+                    tracing::error!("phoenix HTTP service task has already exited");
                 }
-            })
-            .unwrap()
+
+                // This should happen quickly, abort if it takes too long
+                let max_wait = std::time::Instant::now() + std::time::Duration::from_millis(100);
+                let mut interval = tokio::time::interval(std::time::Duration::from_millis(1));
+
+                loop {
+                    if http_task.is_finished() || std::time::Instant::now() > max_wait {
+                        break;
+                    }
+
+                    interval.tick().await;
+                }
+
+                if !http_task.is_finished() {
+                    http_task.abort();
+                }
+
+                if let Err(err) = http_task.await {
+                    if let Ok(panic) = err.try_into_panic() {
+                        let message = panic
+                            .downcast_ref::<String>()
+                            .map(String::as_str)
+                            .or_else(|| panic.downcast_ref::<&str>().copied())
+                            .unwrap_or("<unknown non-string panic>");
+
+                        tracing::error!(panic = message, "phoenix HTTP task paniced");
```

Comment:
Not important, but just fyi, the past tense of `panic` is `panicked`.
```suggestion
                        tracing::error!(panic = message, "phoenix HTTP task panicked");
```